### DISCUSSION TITLE
Sweetening the syntactic sugar for the where clause (FieldsWhere)

### DISCUSSION THREAD

**Discussion by jayasuryat** [vestrel00/contacts-android#162] [Ideas]
The syntactic sugar provided by the [`contacts.core.util.FieldsWhere.kt`](https://github.com/vestrel00/contacts-android/blob/main/core/src/main/java/contacts/core/util/FieldsWhere.kt) can be boiled down into only two functions using Kotlin's `operator fun invoke` overload.

These two methods seem enough to the job so far :
```kotlin
inline operator fun <T : Field> T.invoke(
    where: T.() -> Where<T>,
): Where<T> = where(this)

inline operator fun <T : Field, FS : FieldSet<T>> FS.invoke(
    where: FS.() -> Where<T>,
): Where<T> = where(this)
```

<p>
  <details>
    <summary> Checked against the example given in the Readme file, and it is working fine.</summary>

```kotlin
val contacts = Contacts(context)
    .query()
    .where {
        (Name.GivenName startsWith "leo") and
        (Email.Address { endsWith("gmail.com") or endsWith("hotmail.com") }) and
        (Address.Country equalToIgnoreCase "us") and
        (Event { (Date lessThan Date().toWhereString()) and (Type equalTo EventEntity.Type.BIRTHDAY) }) and
        (Contact.Options.Starred equalTo true) and
        (Nickname.Name equalTo "DarEdEvil") and
        (Organization.Company `in` listOf("facebook", "FB")) and
        (Note.Note.isNotNullOrEmpty())
    }
    .accounts(
        Account("john.doe@gmail.com", "com.google"),
        Account("john.doe@myspace.com", "com.myspace"),
    )
    .include { setOf(
        Contact.Id,
        Contact.DisplayNamePrimary,
        Phone.Number
    ) }
    .orderBy(ContactsFields.DisplayNamePrimary.desc())
    .offset(0)
    .limit(5)
    .find()
```

  </details>
</p>

Would like to know your opinion on this approach, and are there any more cases to consider?

**Comment by vestrel00** [comment]
@JayaSuryaT, I just tried it out... IT IS GENIUS MY FRIEND!!! PURE GENIUS!!!  🤯  🤯  🤯  🤯  🤯  🤯  🤯 

![genius](https://user-images.githubusercontent.com/4332753/147956993-8fed5789-8b6e-453f-9302-844feb489571.gif)

I just tried it out and it works perfectly! I It's so much better than my current implementation because...

1. It is 2 lines of code versus mine which is around 500 lines of code  🤯 !!!
2. It does not require updating when fields/fieldsets are added, modified, or removed.
3. The code autocomplete popup suggestions by Android Studio is much, much cleaner now!

I have reopened issue #150 and assigned it to you. Please **delete the entire  `contacts.core.util.FieldsWhere.kt` file** and add your two functions in `contacts.core.Where` =) 

I will be awarding you a **gold medal** in Reddit once it is merged!!!

> This Reddit comment; https://www.reddit.com/r/Kotlin/comments/rtxfdm/comment/hr19srb/?utm_source=share&utm_medium=web2x&context=3

**Reply by jayasuryat** [reply]
Sure thing @vestrel00, I'll take up the issue and add these methods as top-level methods in the `contacts.core.Where` file.

And thank you so much for your kind words, I really appreciate you taking the time to express that.


<img src="https://tenor.com/view/you-just-made-my-day-tupac-demetrius-shipp-jr-all-eyez-on-me-thank-you-gif-18331997.gif" width="249" height="117" />

**Comment by vestrel00** [comment]
@JayaSuryaT, I merged your PR! I also left a comment on it; https://github.com/vestrel00/contacts-android/pull/163#discussion_r777780600

<img width="789" alt="Screen Shot 2022-01-03 at 3 35 14 PM" src="https://user-images.githubusercontent.com/4332753/147998199-f342a54e-966d-45cc-802a-347f013a1082.png">

Thanks again =)
